enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,243 @@
+1. Required configuration for supporting Data Disk should be added to the `AzureMachineProviderSpec`. 
+This will allow users to set Ultra Disks as Data Disks for the Machine. This feature has been made available upstream in Cluster API Provider for Azure (CAPZ) by: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1478
+1. Required configuration for specifying the ability in `AzureMachineProviderSpec` to allow the Azure CSI driver to provision Ultra Disks as PVs for the Machine. See the issue tracking the implementation of this feature in the upstream Cluster API Provider for Azure (CAPZ): https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1852
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `*ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  NameSuffix string `json:"nameSuffix"`
```

Comment:
Are there any limitations on characters or length for this field? Is it required?
We can specify both patterns and required/optional via kubebuilder tags to make it obvious what we are expecting

Reply:
That's a good point 👍 
Looks like in upstream the field is implicitly required as if that's not present the validation fails with an error. And there is also a check on uniqueness: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2349883d1effe6344690459b40e5dc26138010df/api/v1beta1/azuremachine_validation.go#L118-L126

One thing that is missing in the upstream validation is that the overall Data Disk Name (not just the `NameSuffix`) must be at most 80 chars long, and that there is a specific validation pattern (see img below).

The problem I see here with doing this validation in Kubebuilder is that we don't have control over the overall length of the name, as the `<machineName>_` will be prepended to the string after the Kubebuilder validation has happened.

<img width="752" alt="Screenshot 2022-01-28 at 16 37 00" src="https://user-images.githubusercontent.com/4902157/151592658-bb35dd3e-cc2f-4c60-823b-cd0b7421ac46.png">

Reply:
Two potential alternatives to having the Kubebuilder validation tags, could be:

1) Do the validation when the NameSuffix and the Machine name are composed:
    ```
	    for i, disk := range vmSpec.DataDisks {
		    dataDisks[i] = compute.DataDisk{
			    CreateOption: compute.DiskCreateOptionTypesEmpty,
			    DiskSizeGB:   to.Int32Ptr(disk.DiskSizeGB),
			    Lun:          disk.Lun,
			    Name:         to.StringPtr(azure.GenerateDataDiskName(vmSpec.Name, disk.NameSuffix)), // <--
			    Caching:      compute.CachingTypes(disk.CachingType),
		    }
		    ...
	    }
	    
	    ...
	    // GenerateDataDiskName generates OS disk name used by VM
	    func GenerateDataDiskName(machineName, suffix string) string {
	            // ###  the composite string could be validated here ###
		    return fmt.Sprintf("%s_%s", machineName, suffix)
	    }
    ```

2) Just leave it to the Provider API to validate and return the error.
    
    invalid nameSuffix field:
    ```
    spec:
      providerSpec:
        value:
          apiVersion: machine.openshift.io/v1beta1
          kind: AzureMachineProviderSpec
          dataDisks:
          - nameSuffix: testdiskkjfoifhdsgfsdg08wehflkdhlfsdlkfdslkfslksfnsdlbsjbgfkjgbkjbdkasbkjsabaskjbdakjdbakjdbasdkajsbdaskjbdaksjbdaskjbdaskjbdksajdbaskjdbkjasbdaksjbdkasjbd
            diskSizeGB: 4
            managedDisk:
              storageAccountType: UltraSSD_LRS
            lun: 0
    ```
    error that shows up in the controller logs + machine in "Failed" state:
    ```
    I0131 17:45:00.384762       1 virtualmachines.go:126] creating vm ddonati-test15-kdq8r-worker3-eastus3-npm8d
    I0131 17:45:04.267389       1 machine_scope.go:192] ddonati-test15-kdq8r-worker3-eastus3-npm8d: patching machine
    E0131 17:45:04.307090       1 actuator.go:78] Machine error: failed to reconcile machine "ddonati-test15-kdq8r-worker3-eastus3-npm8d": failed to create vm ddonati-test15-kdq8r-worker3-eastus3-npm8d: failure sending request for machine ddonati-test15-kdq8r-worker3-eastus3-npm8d: cannot create vm: compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="The entity name 'dataDisk.name' is invalid according to its validation rule: ^[^_\\W][\\w-._]{0,79}(?<![-.])$." Target="dataDisk.name"
    W0131 17:45:04.307113       1 controller.go:388] ddonati-test15-kdq8r-worker3-eastus3-npm8d: failed to create machine: failed to reconcile machine "ddonati-test15-kdq8r-worker3-eastus3-npm8d": failed to create vm ddonati-test15-kdq8r-worker3-eastus3-npm8d: failure sending request for machine ddonati-test15-kdq8r-worker3-eastus3-npm8d: cannot create vm: compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="The entity name 'dataDisk.name' is invalid according to its validation rule: ^[^_\\W][\\w-._]{0,79}(?<![-.])$." Target="dataDisk.name"
    I0131 17:45:04.307126       1 controller.go:504] ddonati-test15-kdq8r-worker3-eastus3-npm8d: going into phase "Failed"
    ```

Reply:
Option 2 leaves a rather complicated error message for the customer to interpret, I think we can do better than that and provide them a nicer error by going with option 1

Reply:
Agree. I'll add it to the implementation.

Reply:
Done in https://github.com/openshift/enhancements/pull/1021/commits/044f4e93d9a2697c16dbb4299dc8fa0a90560a11